File: forge-core/internal/platformcorecontract/artifact.go
package platformcorecontract

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"
)

// ValidateArtifactRef validates supplied ArtifactRef declarations only. Composed
// contracts may request exactly one ordered stage: values, references, or relations.
func ValidateArtifactRef(value *ArtifactRef, stages ...string) error {
	if len(stages) == 0 {
		_, err := CanonicalArtifactRefJSON(value)
		return err
	}
	if len(stages) != 1 || value == nil {
		return reject(rejectionDocumentInvalid, "one non-nil staged ArtifactRef is required")
	}
	switch stages[0] {
	case "values":
		return withRejection(validateArtifactValues(value), rejectionValueInvalid)
	case "references":
		return validateArtifactReferences(value)
	case "relations":
		return validateArtifactRelations(value)
	default:
		return reject(rejectionValueInvalid, "ArtifactRef validation stage is unsupported")
	}
}

// CanonicalArtifactRefJSON returns exact compact canonical v1 bytes.
func CanonicalArtifactRefJSON(value *ArtifactRef) ([]byte, error) {
	if err := validateArtifactFields(value); err != nil {
		return nil, withRejection(err, rejectionValueInvalid)
	}
	canonical, err := typedCanonical(value, maxArtifactRefBytes)
	return canonical, withRejection(err, rejectionValueInvalid)
}

// DecodeCanonicalArtifactRef accepts only one exact canonical v1 ArtifactRef.
func DecodeCanonicalArtifactRef(data []byte) (*ArtifactRef, error) {
	var value ArtifactRef
	if err := decodeTypedCanonical(data, maxArtifactRefBytes, &value); err != nil {
		return nil, withRejection(err, rejectionDocumentInvalid)
	}
	if err := validateExactTypedDocument(data, &value, maxArtifactRefBytes, "ArtifactRef"); err != nil {
		return nil, err
	}
	if err := validateArtifactFields(&value); err != nil {
		return nil, withRejection(err, rejectionValueInvalid)
	}
	return &value, nil
}

// ArtifactRefSHA256 returns a domain-separated conformance observation digest.
func ArtifactRefSHA256(value *ArtifactRef) (string, error) {
	canonical, err := CanonicalArtifactRefJSON(value)
	if err != nil {
		return "", err
	}
	return observationDigest(artifactDigestDomain, canonical), nil
}

func validateArtifactFields(value *ArtifactRef) error {
	if value == nil {
		return reject(rejectionDocumentInvalid, "ArtifactRef is required")
	}
	if err := validateArtifactValues(value); err != nil {
		return err
	}
	if err := validateArtifactReferences(value); err != nil {
		return err
	}
	return validateArtifactRelations(value)
}

func validateArtifactValues(value *ArtifactRef) error {
	if value.Canonicalization != CanonicalizationV1 {
		return fmt.Errorf("ArtifactRef canonicalization is unsupported")
	}
	if err := validateLowerToken(value.ArtifactKind, "artifact_kind", 64); err != nil {
		return err
	}
	if err := validateHash(value.ContentDigest, "content_digest"); err != nil {
		return err
	}
	if err := validateTypedID(value.LogicalID, "art", "logical_id"); err != nil {
		return err
	}
	return validateArtifactDetails(value)
}

func validateArtifactDetails(value *ArtifactRef) error {
	if err := validateMediaType(value.MediaType); err != nil {
		return err
	}
	if err := validateTypedID(value.ProducerAttemptID, "atm", "producer_attempt_id"); err != nil {
		return err
	}
	if err := validateEntityRef(value.SourceSnapshotRef, "source_snapshot_ref"); err != nil {
		return err
	}
	if err := validateRecordRef(value.ProvenanceRef, "provenance_ref"); err != nil {
		return err
	}
	return validateArtifactClassification(value)
}

func validateArtifactReferences(value *ArtifactRef) error {
	if value.SourceSnapshotRef.EntityType != entityProjectSnapshot {
		return reject(rejectionReferenceMismatch, "source_snapshot_ref must reference project_snapshot")
	}
	return nil
}

func validateArtifactRelations(value *ArtifactRef) error {
	if value.ContentID != "sha256:"+value.ContentDigest {
		return reject(rejectionRelationMismatch, "content_id must equal sha256: plus content_digest")
	}
	return nil
}

func validateArtifactClassification(value *ArtifactRef) error {
	if err := validateUnixMS(value.CreatedAtUnixMS, "created_at_unix_ms"); err != nil {
		return err
	}
	if value.SizeBytes < 0 || value.SizeBytes > maxArtifactBytes {
		return fmt.Errorf("size_bytes must be in 0..%d", maxArtifactBytes)
	}
	if !oneOf(value.RetentionClass, "audit", "durable", "ephemeral", "legal_hold", "project") {
		return fmt.Errorf("retention_class is unsupported")
	}
	if !oneOf(value.Sensitivity, "confidential", "internal", "public", "secret") {
		return fmt.Errorf("sensitivity is unsupported")
	}
	return nil
}

func validateMediaType(value string) error {
	if err := validateText(value, "media_type", 128, true); err != nil {
		return err
	}
	parts := strings.Split(value, "/")
	if len(parts) != 2 || parts[0] == "" || parts[1] == "" {
		return fmt.Errorf("media_type must contain one type/subtype separator")
	}
	for _, part := range parts {
		if !lowerMediaTypeByte(part[0]) {
			return fmt.Errorf("media_type type and subtype must start with lowercase alphanumeric")
		}
		for _, character := range []byte(part) {
			if !mediaTypeByte(character) {
				return fmt.Errorf("media_type contains unsupported byte")
			}
		}
	}
	return nil
}

func lowerMediaTypeByte(value byte) bool {
	return value >= 'a' && value <= 'z' || value >= '0' && value <= '9'
}

func mediaTypeByte(value byte) bool {
	return value >= 'a' && value <= 'z' || value >= '0' && value <= '9' ||
		strings.ContainsRune("!#$&^_.+-", rune(value))
}

func observationDigest(domain string, canonical []byte) string {
	hasher := sha256.New()
	_, _ = hasher.Write([]byte(domain))
	_, _ = hasher.Write(canonical)
	return hex.EncodeToString(hasher.Sum(nil))
}

func oneOf(value string, options ...string) bool {
	for _, option := range options {
		if value == option {
			return true
		}
	}
	return false
}

File: forge-runtime/crates/domain/src/platform_core_contract/artifact.rs
use super::{
    ArtifactRef, CANONICALIZATION, EntityType, MAX_ARTIFACT_BYTES, MAX_UNIX_MILLISECONDS,
    PlatformCoreContractError, RejectionCode, RetentionClass, Sensitivity, identity, invalid,
    references, reject, wire,
};

/// Validates supplied `ArtifactRef` declarations without resolving content.
///
/// # Errors
///
/// Returns an error for malformed identity, provenance, classification, or bounds.
pub fn validate_artifact_ref(value: &ArtifactRef) -> Result<(), PlatformCoreContractError> {
    validate_artifact_values(value)?;
    validate_artifact_references(value)?;
    validate_artifact_relations(value)
}

pub(super) fn validate_artifact_values(
    value: &ArtifactRef,
) -> Result<(), PlatformCoreContractError> {
    if value.canonicalization != CANONICALIZATION {
        return Err(invalid("ArtifactRef canonicalization is unsupported"));
    }
    if !wire::lower_token(&value.artifact_kind) || value.artifact_kind.len() > 64 {
        return Err(invalid("artifact_kind must be a bounded lowercase token"));
    }
    wire::validate_hash(&value.content_digest, "content_digest")?;
    identity::validate_typed_id(&value.logical_id, "art", "logical_id")?;
    validate_artifact_details(value)
}

fn validate_artifact_details(value: &ArtifactRef) -> Result<(), PlatformCoreContractError> {
    validate_media_type(&value.media_type)?;
    identity::validate_typed_id(&value.producer_attempt_id, "atm", "producer_attempt_id")?;
    references::validate_entity_ref(&value.source_snapshot_ref, "source_snapshot_ref")?;
    references::validate_record_ref(&value.provenance_ref, "provenance_ref")?;
    validate_artifact_bounds(value)
}

pub(super) fn validate_artifact_references(
    value: &ArtifactRef,
) -> Result<(), PlatformCoreContractError> {
    if value.source_snapshot_ref.entity_type != EntityType::ProjectSnapshot {
        return Err(reject(
            RejectionCode::ReferenceMismatch,
            "source_snapshot_ref must reference project_snapshot",
        ));
    }
    Ok(())
}

pub(super) fn validate_artifact_relations(
    value: &ArtifactRef,
) -> Result<(), PlatformCoreContractError> {
    if value.content_id != format!("sha256:{}", value.content_digest) {
        return Err(reject(
            RejectionCode::RelationMismatch,
            "content_id must equal sha256: plus content_digest",
        ));
    }
    Ok(())
}

fn validate_artifact_bounds(value: &ArtifactRef) -> Result<(), PlatformCoreContractError> {
    if !(1..=MAX_UNIX_MILLISECONDS).contains(&value.created_at_unix_ms) {
        return Err(invalid(format!(
            "created_at_unix_ms must be in 1..{MAX_UNIX_MILLISECONDS}"
        )));
    }
    if !(0..=MAX_ARTIFACT_BYTES).contains(&value.size_bytes) {
        return Err(invalid(format!(
            "size_bytes must be in 0..{MAX_ARTIFACT_BYTES}"
        )));
    }
    if matches!(&value.retention_class, RetentionClass::Unknown(_)) {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "retention_class is unsupported",
        ));
    }
    if matches!(&value.sensitivity, Sensitivity::Unknown(_)) {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "sensitivity is unsupported",
        ));
    }
    Ok(())
}

fn validate_media_type(value: &str) -> Result<(), PlatformCoreContractError> {
    wire::validate_text(value, "media_type", 128, true)?;
    let parts: Vec<&str> = value.split('/').collect();
    if parts.len() != 2 || parts.iter().any(|part| part.is_empty()) {
        return Err(invalid(
            "media_type must contain one type/subtype separator",
        ));
    }
    if parts.iter().any(|part| {
        !part.as_bytes()[0].is_ascii_lowercase() && !part.as_bytes()[0].is_ascii_digit()
    }) {
        return Err(invalid(
            "media_type type and subtype must start with lowercase alphanumeric",
        ));
    }
    if parts
        .iter()
        .flat_map(|part| part.bytes())
        .any(|byte| !media_type_byte(byte))
    {
        return Err(invalid("media_type contains an unsupported byte"));
    }
    Ok(())
}

fn media_type_byte(value: u8) -> bool {
    value.is_ascii_lowercase()
        || value.is_ascii_digit()
        || matches!(
            value,
            b'!' | b'#' | b'$' | b'&' | b'^' | b'_' | b'.' | b'+' | b'-'
        )
}
